src/main/java/net/snowflake/client/core/SFArrowResultSet.java

```diff
@@ -564,32 +565,81 @@ public Timestamp getTimestamp(int columnIndex, TimeZone tz) throws SFException {
-    if (type == SnowflakeUtil.EXTRA_TYPES_VECTOR) {
+    int columnType = resultSetMetaData.getColumnType(columnIndex);
+    if (columnType == SnowflakeUtil.EXTRA_TYPES_VECTOR) {
       return getString(columnIndex);
     }
-    ArrowVectorConverter converter = currentChunkIterator.getCurrentConverter(columnIndex - 1);
+
+    ArrowVectorConverter converter = getConfiguredConverter(columnIndex);
     int index = currentChunkIterator.getCurrentRowInRecordBatch();
     wasNull = converter.isNull(index);
-    converter.setTreatNTZAsUTC(treatNTZAsUTC);
-    converter.setUseSessionTimezone(useSessionTimezone);
-    converter.setSessionTimeZone(sessionTimeZone);
-    Object obj = converter.toObject(index);
     boolean isStructuredType = resultSetMetaData.isStructuredTypeColumn(columnIndex);
-    if (isVarcharConvertedStruct(type, isStructuredType, converter)) {
-      if (obj != null) {
+    Object obj = converter.toObject(index);
+    if (obj == null) {
+      return null;
+    }
+    if (columnType == Types.STRUCT && isStructuredType) {
+      if (converter instanceof VarCharConverter) {
         return new StructObjectWrapper((String) obj, createJsonSqlInput(columnIndex, obj));
+      } else if (converter instanceof StructConverter) {
+        return new StructObjectWrapper(
+            converter.toString(index),
+            createArrowSqlInput(columnIndex, (Map<String, Object>) obj),
+            obj);
       }
     }
-    return obj;
+    return new StructObjectWrapper(converter.toString(index), null, obj);
   }
 
-  private boolean isVarcharConvertedStruct(
-      int type, boolean isStructuredType, ArrowVectorConverter converter) {
-    return type == Types.STRUCT && isStructuredType && converter instanceof VarCharConverter;
+  @Override
+  public <T> Object getObject(int columnIndex, Class<T> type) throws SFException {
+    if (String.class.isAssignableFrom(type)) {
+      return getObject(columnIndex);
+    }
+
+    int columnType = resultSetMetaData.getColumnType(columnIndex);
+    if (columnType == SnowflakeUtil.EXTRA_TYPES_VECTOR) {
+      return getString(columnIndex);
+    }
+    ArrowVectorConverter converter = getConfiguredConverter(columnIndex);
+    int index = currentChunkIterator.getCurrentRowInRecordBatch();
+    wasNull = converter.isNull(index);
+    Object obj = converter.toObject(index);
+    boolean isStructuredType = resultSetMetaData.isStructuredTypeColumn(columnIndex);
+    if (obj == null) {
+      return null;
+    }
+    if (columnType == Types.STRUCT && isStructuredType) {
+      if (converter instanceof VarCharConverter) {
+        return new StructObjectWrapper(null, createJsonSqlInput(columnIndex, obj));
+      } else if (converter instanceof StructConverter) {
+        return new StructObjectWrapper(
+            null, createArrowSqlInput(columnIndex, (Map<String, Object>) obj));
+      }
+    }
+    return new StructObjectWrapper(null, null, obj);
+  }
+
+  @SnowflakeJdbcInternalApi
```

Comment:
private method does not need the internal annotation

Reply:
good catch, I'll remove